src/v/model/record_batch_reader.cc

```diff
@@ -29,6 +29,11 @@ using data_t = record_batch_reader::data_t;
 using foreign_data_t = record_batch_reader::foreign_data_t;
 using storage_t = record_batch_reader::storage_t;
 
+std::optional<reader_data_layout>
+maybe_get_data_layout(const model::record_batch_reader& reader) {
+    return reader._impl->maybe_get_data_layout();
+}
+
```

Comment:
> only works with in-memory readers

can we factor this out into a consumer that cloud_topics owns which computes the layout / stats (e.g. something like https://github.com/redpanda-data/redpanda/pull/23272/files#diff-ef46c22533ae719434fb4042a1aab6ece44087845e14ffad8022bf66227dec50R43)? i'd be in favor of adding this to the record batch reader once we supported different backends beyond in-memory.

is there a specific reason why we added to the reader interface?

Reply:
I wanted to avoid pulling everything into memory. Maybe it doesn't make much sense since the incoming record batch readers are also stored in memory.

Reply:
it only works for in-memory readers right now. adding the consumer version won't add any additional overhead.